File: solution-python/020_valid_parentheses.py
"""
	020. Valid Parentheses
	@name: li jin
	@date: Mar 02, 2016
	@link: https://leetcode.com/problems/valid-parentheses/
	@time: 40 ms
"""

class Solution(object):
    def isValid(self, s):
        lp = {'(', '{', '['}
        rp = {')' : '(', '}' : '{', ']' : '['}
        stack = []
        for c in s:
            if c in lp:
                stack.append(c)
            elif not stack or stack[-1] != rp[c]:
                return False
            else:
                stack.pop(-1)
        return len(stack) == 0

File: solution-cpp/020_valid_parentheses.cpp
/*
	020. Valid Parentheses
	@name: li jin
	@date: Mar 02, 2016
	@link: https://leetcode.com/problems/valid-parentheses/
	@time: 0 ms
*/

class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        map<char, char> match;
        match['}'] = '{'; match[']'] = '['; match[')'] = '(';
        for (auto c : s) {
            if (match.find(c) != match.end()) {
                if (st.empty() || st.top() != match[c]) return false;
                else st.pop();
            }
            else {
                st.push(c);
            }
        }
        return st.empty();
    }
};
